Log when an insecure Https -> Http redirect is blocked

rmkerr opened the pull request:
.NET Core does not allow insecure redirects from https to http, but .NET Framework does. Developers have had some trouble diagnosing the issue when they run into this difference. This change adds logging to help developers track down what is going on.

Fixes: #23813 (was https://github.com/dotnet/corefx/issues/24577)

Commit:
Add logging for insecure redirects on WinHttp.

Commit:
Improve WinHttp logging, add CurlHandler logging.

davidsh commented on src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:
```diff
@@ -888,6 +888,16 @@ private void EnsureSessionHandleExists(WinHttpRequestState state)
                 HttpResponseMessage responseMessage = WinHttpResponseParser.CreateResponseMessage(state, _doManualDecompressionCheck);
                 state.Tcs.TrySetResult(responseMessage);
+
+                if((responseMessage.StatusCode == HttpStatusCode.MultipleChoices ||
```
nit: add space after 'if'

davidsh commented on src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs:
```diff
@@ -370,6 +370,10 @@ internal void SetPossibleRedirectForLocationHeader(string location)
                         SetCookieOption(newUri);
                     }
+
+                    if(newUri.Scheme == Uri.UriSchemeHttp && _requestMessage.RequestUri.Scheme == Uri.UriSchemeHttps) {
```
nit: add space after 'if'

davidsh replied:
nit: bracing style should match rest of file, i.e. Allman-style

davidsh commented on src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:
```diff
@@ -888,6 +888,16 @@ private void EnsureSessionHandleExists(WinHttpRequestState state)
+                    responseMessage.StatusCode == HttpStatusCode.RedirectKeepVerb) &&
+                   state.RequestMessage.RequestUri.Scheme == Uri.UriSchemeHttps &&
+                   responseMessage.Headers.Location?.Scheme == Uri.UriSchemeHttp) {
+                    WinHttpTraceHelper.Trace("WinHttpHandler.SendAsync: Insecure https to http redirect blocked.");
```
It would be nice if we had EventSourceTrace style tracing available in WinHttpHandler (similar to CurlHandler, etc.), but that is probably beyond the scope of this PR.

davidsh commented on src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:
```diff
@@ -888,6 +888,16 @@ private void EnsureSessionHandleExists(WinHttpRequestState state)
+                    responseMessage.StatusCode == HttpStatusCode.RedirectMethod ||
+                    responseMessage.StatusCode == HttpStatusCode.RedirectKeepVerb) &&
+                   state.RequestMessage.RequestUri.Scheme == Uri.UriSchemeHttps &&
+                   responseMessage.Headers.Location?.Scheme == Uri.UriSchemeHttp) {
```
nit: bracing style should match rest of file, i.e. Allman-style

danmosemsft replied:
Are the urls in the payload or should they be included in the trace message?

rmkerr replied:
They're not in the payload. so I'll add it to the message. What do you think of wording it something like this?
```
$"Insecure https to http redirect from {RequestUri} to {DestinationUri} blocked."
```

caesar1995 commented on src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:
```diff
@@ -888,6 +888,16 @@ private void EnsureSessionHandleExists(WinHttpRequestState state)
+                    responseMessage.StatusCode == HttpStatusCode.MovedPermanently ||
+                    responseMessage.StatusCode == HttpStatusCode.Redirect ||
+                    responseMessage.StatusCode == HttpStatusCode.RedirectMethod ||
+                    responseMessage.StatusCode == HttpStatusCode.RedirectKeepVerb) &&
```
Should we consider status code 308: Permanent Redirect here?

davidsh replied:
Maybe it would be less code/higher perf to just check for any 3xx code, i.e. do some simple math operation than listing all these out.

rmkerr replied:
I considered that but was concerned about `304 Not Modified` responses, which have a different behavior.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs:
```diff
@@ -370,6 +370,10 @@ internal void SetPossibleRedirectForLocationHeader(string location)
                         SetCookieOption(newUri);
                     }
+
+                    if(newUri.Scheme == Uri.UriSchemeHttp && _requestMessage.RequestUri.Scheme == Uri.UriSchemeHttps) {
```
Interestingly we could use this same check to avoid doing some of the above work if we're not going to redirect anyway.  But it's probably not something worth optimizing for.

stephentoub commented:
Can you add similar tracing to SocketsHttpHandler?  Probably would go here:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticateAndRedirectHandler.cs#L135

rmkerr commented:
Thanks for the feedback! I'll send out another iteration with the minor changes and with SocketsHttpHandler included.

Commit:
Add SocketsHttpHandler logging, add source and dest URIs to logging.

rmkerr commented on src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:
```diff
@@ -888,6 +888,14 @@ private void EnsureSessionHandleExists(WinHttpRequestState state)
+
+                if (((responseMessage.StatusCode >= HttpStatusCode.MultipleChoices && responseMessage.StatusCode <= HttpStatusCode.SeeOther) ||
+                     (responseMessage.StatusCode >= HttpStatusCode.RedirectKeepVerb && responseMessage.StatusCode <= HttpStatusCode.PermanentRedirect)) &&
+                    state.RequestMessage.RequestUri.Scheme == Uri.UriSchemeHttps && responseMessage.Headers.Location?.Scheme == Uri.UriSchemeHttp)
```
This whole block is pretty ugly, but in most cases it will make fewer checks than it would before. It also covers the case of `HttpStatusCode.PermanentRedirect`, which it did not previously.

rmkerr commented:
I've addressed review comments and added logging to `SocketsHttpHandler`. While doing that, I did notice that `SocketsHttpHandler` has extremely limited logging compared to `WinHttpHandler` and 'CurlHandler`. I don't think that is within the scope of this PR, but I think that it is worth looking into later.

Commit:
WinHttpHandler builds against Net Standard, which does not include new status codes like 308. Instead of using PermanentRedirect, we have to cast the integer 308 to an HttpStatusCode.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs:
```diff
@@ -370,6 +370,11 @@ internal void SetPossibleRedirectForLocationHeader(string location)
+
+                    if (newUri.Scheme == Uri.UriSchemeHttp && _requestMessage.RequestUri.Scheme == Uri.UriSchemeHttps)
+                    {
+                        EventSourceTrace($"Insecure https to http redirect from {_requestMessage.RequestUri} to {newUri} blocked.");
```
This is unfortunately a relatively expensive operation.  This is translated by the compiler into:
```C#
EventSourceTrace(string.Format("Insecure https to http redirect from {0} to {1} blocked.", new object[] { _requestMessage.RequestUri, newUri }));
```
so even if tracing is disabled, we still end up allocating that object[], and we still end up doing the string format.

It should instead do something like:
https://github.com/rmkerr/corefx/blob/c1a2eddf4ee509b1e591f8e63ad7429625e42b20/src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs#L190
where the argument is passed in as an argument, and all of the work is deferred until in the EventSourceTrace method it checks for whether tracing is enabled.  I don't recall for sure, but I don't think there's currently an EventSourceTrace overload that takes two arguments to be substituted; assuming not, you could just use a value tuple, e.g.
```C#
EventSourceTrace("Insecure https to http redirect: {0}", (_requestMessage.RequestUri, newUri));
```

rmkerr replied:
I'm used to thinking of that kind of string operation as "free", but obviously that isn't the case with format strings. I'll make the change, and the cost of format strings is something I'll keep in mind!

stephentoub commented on src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:
```diff
@@ -888,6 +888,15 @@ private void EnsureSessionHandleExists(WinHttpRequestState state)
+                    state.RequestMessage.RequestUri.Scheme == Uri.UriSchemeHttps && responseMessage.Headers.Location?.Scheme == Uri.UriSchemeHttp)
+                {
+                    WinHttpTraceHelper.Trace("WinHttpHandler.SendAsync: Insecure https to http redirect from" +
+                        $"{state.RequestMessage.RequestUri} to {responseMessage.Headers.Location} blocked.");
```
See my comment on the CurlHandler change.  There are similar costs here.  More, actually, as this will create an extra string for concatenating the const string with the result of the string interpolation.

rmkerr replied:
This should be mostly resolved. I can save the cost of the call to format, but do still incur the cost of casting both URIs to strings. I could fix that, but as far as I can tell I would have to modify WinHttpTraceHelper to do that.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticateAndRedirectHandler.cs:
```diff
@@ -132,6 +132,10 @@ await AuthenticationHelper.TrySetDigestAuthToken(request, currentCredential, dig
                 {
+                    if (NetEventSource.IsEnabled)
+                    {
+                        NetEventSource.Info(this, $"Insecure https to http redirect from {request.RequestUri} to {location} blocked.");
```
Unlike the other two, this one is fine, as it's guarded by the NetEventSource.IsEnabled check.  The costs will only be incurred when tracing is enabled.

stephentoub commented:
> While doing that, I did notice that SocketsHttpHandler has extremely limited logging compared to WinHttpHandler and CurlHandler.

That's true in that auth/redirect file.  It could certainly stand to be improved.  That said, @geoffkizer is currently doing some refactoring here, so I'd recommend holding off on significant changes there until his changes land.

Commit:
Ensure that String.Format is not called when logging is disabled.

davidsh commented on src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:
```diff
@@ -888,6 +888,15 @@ private void EnsureSessionHandleExists(WinHttpRequestState state)
+                     (responseMessage.StatusCode >= HttpStatusCode.RedirectKeepVerb && responseMessage.StatusCode <= (HttpStatusCode)308)) &&
+                    state.RequestMessage.RequestUri.Scheme == Uri.UriSchemeHttps && responseMessage.Headers.Location?.Scheme == Uri.UriSchemeHttp)
+                {
+                    WinHttpTraceHelper.Trace("WinHttpHandler.SendAsync: Insecure https to http redirect from {0} to {1} blocked.",
```
If you need to wrap this so that this doesn't cause extra work if tracing isn't enabled, then wrap this in a if (WinHttpTraceHelper.IsEnabled) etc.

rmkerr replied:
I'll make that change. Even though it means making that truly horrible if-statement even worse.

pjanotti commented on src/System.Net.Http/src/System/Net/Http/CurlHandler/CurlHandler.EasyRequest.cs:
```diff
@@ -370,6 +370,11 @@ internal void SetPossibleRedirectForLocationHeader(string location)
+
+                    if (newUri.Scheme == Uri.UriSchemeHttp && _requestMessage.RequestUri.Scheme == Uri.UriSchemeHttps)
+                    {
+                        EventSourceTrace($"Insecure https to http redirect: {0}", (_requestMessage.RequestUri, newUri));
```
Remove the interpolation ```$```

rmkerr replied:
I missed that you made this comment! I'll try to address that.

Commit:
Check if tracing is enabled in WInHttp before testing the status code and schemes. Avoids extra work in the case that logging is not enabled.

Commit:
Remove interpolation

pjanotti reviewed (approved):
:shipit:

rmkerr commented:
CI is green. @stephentoub @davidsh any additional comments, or are you okay with me going ahead and merging this?

stephentoub commented:
LGTM.  FYI this is going to conflict with @geoffkizer's change at https://github.com/dotnet/corefx/pull/27266.

rmkerr commented:
Hey @geoffkizer! Would you prefer if I held off on merging this until your PR is through? I think it should be easy to reconcile the changes either way.
